addons/stock/migrations/12.0.1.1/post-migration.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2018 Eficent <http://www.eficent.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo import _
+from openupgradelib import openupgrade, openupgrade_merge_records
+from psycopg2.extensions import AsIs
+
+
+def fill_stock_picking_type_barcode(env):
+    picking_types = env['stock.picking.type'].with_context(
+        active_test=False).search([('code', '!=', False),
+                                   (['warehouse_id', '!=', False])])
+    default_names_dict = {
+        _('Receipts'): '-RECEIPTS',
+        _('Delivery Orders'): '-DELIVERY',
+        _('Pack'): '-PACK',
+        _('Pick'): '-PICK',
+        _('Internal Transfers'): '-INTERNAL',
+    }
+    for picking_type in picking_types:
+        if picking_type.name in default_names_dict:
+            picking_type.barcode = picking_type.warehouse_id.code.replace(
+                " ", "").upper() + default_names_dict[picking_type.name]
+
+
+def merge_stock_location_path_stock_rule(env):
+    openupgrade.logged_query(
+        env.cr, """
+        INSERT INTO stock_rule (name, active, action, sequence, company_id,
+            location_id, location_src_id, route_id, procure_method,
+            route_sequence, picking_type_id, delay, propagate, warehouse_id,
+            auto, create_uid, create_date, write_uid, write_date, %s)
+        SELECT name, active, 'push' AS action, sequence, company_id,
+            location_id, location_src_id, route_id,
+            'make_to_stock' AS procure_method, route_sequence,
+            picking_type_id, delay, propagate, warehouse_id, auto,
+            create_uid, create_date, write_uid, write_date, id
+        FROM stock_location_path
+        """, (AsIs(openupgrade.get_legacy_name('loc_path_id')), ),
+    )
+    openupgrade.logged_query(
+        env.cr, """
+        UPDATE ir_model_data imd
+        SET model = 'stock.rule', res_id = sr.id
+        FROM stock_rule sr
+        WHERE imd.res_id = sr.%s AND model = 'stock.location.path'
+        """, (AsIs(openupgrade.get_legacy_name('loc_path_id')), ),
+    )
+    env.cr.execute(
```

Comment:
Why are you performing this until the end of the method? I don't think it's needed.

Reply:
I think it's needed because the stock_move table has rule_id and push_rule_id columns. So as I add the push rules with a different id in the stock rule table, then I need to make sure that the rule_id column of stock_move is properly filled in two cases: (1) if rule_id and push_rule_id both were filled, and (2) rule_id not filled but push_rule_id filled.